x/market/handler/endblock_test.go

```diff
@@ -0,0 +1,213 @@
+	winner, err := pickBidWinner(w.bids)
+	if !errors.Is(err, w.expErr) {
+		t.Errorf("returned err: %v does not match %v", err, w.expErr)
+	}
+	if w.expWinner != nil && !winner.Equals(w.expWinner.ID()) {
+		t.Errorf("unexpected winner: %#v\n%q : %v", winner, types.BidIDString(winner.BidID), winner.Price)
+		t.Logf("winner: %+v coin: %s", winner.ID(), winner.Price.String())
+	}
+}
+
+func TestBidWinner(t *testing.T) {
+	originOID := testutil.OrderID(t)
+
+	bid0 := createBid(t, originOID, int64(3))
+	bid1 := createBid(t, originOID, int64(5))
+	bid2 := createBid(t, originOID, int64(5))
+	bid3 := createBid(t, originOID, int64(4))
+	bid4 := createBid(t, originOID, int64(5))
+	bid5 := createBid(t, originOID, int64(5))
+	bid6 := createBid(t, originOID, int64(5))
+	bid7 := createBid(t, originOID, int64(8))
+	bid8 := createBid(t, originOID, int64(6))
+	bid9 := createBid(t, originOID, int64(5))
+
+	var winnerTests = []winnerTest{
+		{
+			desc:      "no bids",
+			bids:      []types.Bid{},
+			expWinner: nil,
+			expErr:    errNoBids,
+		},
+		{
+			desc:      "single bid",
+			bids:      []types.Bid{bid9},
+			expWinner: &bid9,
+		},
+		{
+			desc:      "two bids",
+			bids:      []types.Bid{bid7, bid4},
+			expWinner: &bid4,
+		},
+		{
+			desc:      "two matching bids",
+			bids:      []types.Bid{bid9, bid7, bid4},
+			expWinner: &bid4,
+		},
+		{
+			desc:      "all the same bid values",
+			bids:      []types.Bid{bid9, bid9, bid9, bid9, bid9},
+			expWinner: &bid9,
+		},
+		{
+			desc:      "two of the same",
+			bids:      []types.Bid{bid9, bid3},
+			expWinner: &bid3,
+		},
+		{
+			desc:      "multiple of the same, but one lowest value",
+			bids:      []types.Bid{bid1, bid0, bid9, bid2, bid3, bid4},
+			expWinner: &bid0,
+		},
+		{
+			desc:      "confirm last bid in list is picked",
+			bids:      []types.Bid{bid1, bid9, bid2, bid3, bid4, bid0},
+			expWinner: &bid0,
+		},
+		{
+			desc:      "two matching low bids",
+			bids:      []types.Bid{bid5, bid6, bid7, bid8},
+			expWinner: &bid6,
+		},
+	}
+
+	for _, test := range winnerTests {
+		t.Run(test.desc, test.testFunc)
+	}
+}
+
+var errExpUnevenDistribution = errors.New("this test expects uneven winning distribution")
+
+type testDist struct {
+	desc   string
+	bidNum int
+	rounds int
+	expErr error
+}
+
+func (td *testDist) testFunc(t *testing.T) {
+	originOID := testutil.OrderID(t)
+
+	distributionSpread := make(map[int]int)
+	for i := 0; i < td.bidNum; i++ {
+		distributionSpread[i] = 0
+	}
+
+	for i := 0; i < td.rounds; i++ {
+		winCount := make(map[int]int)
+		bIndex := make(map[string]int)
+		bids := make([]types.Bid, 0)
+		// generate N bids all with the same bidding amount
+		for j := 0; j < td.bidNum; j++ {
+			b := createBid(t, originOID, int64(5))
+			bIndex[b.Provider.String()] = j
+			winCount[j] = 0
+			bids = append(bids, b)
+		}
+
+		winner, err := pickBidWinner(bids)
+		if !errors.Is(err, td.expErr) {
+			if !errors.Is(td.expErr, errExpUnevenDistribution) {
```

Comment:
this is the only type of error that is expected.

Reply:
Simplified to simply assert that the error matches the testDist type's expErr.